README.md
# Java Game Genie Encoder/Decoder Program

## Version Information

    Version 1.1
    Author: John Ratliff
    Copyright (C) 2004 emuWorks http://games.technoplaza.net

## Introduction

Java GGEncoder is an encoding and decoding program for game genie codes.
It can take game genie codes, or their decoded equivalents, and produce the  alternate version.
In other words, you can take game genie codes and turn them into "raw" codes (addresses, values, and sometimes compares), and vice versa.

## Random Information

I wrote this program for several reasons.
The most basic of which is as a tool for ROM hacking.
  
You might say, aren't there several other programs which already do this?
Yes, and no.
uggconv will do all the translations, but has no graphical mode.
ggconvc has a graphical mode, but doesn't do all the translations (it lacks a gameboy/gamegear mode).
uggconv could be compiled for any system, but you still have to do that yourself (except on Windows).
ggconvc will only work in Windows (it's written with Microsoft GUI Libraries).
There are a few other programs that do some other interesting things that Java GGEncoder doesn't do.
There is a game gear program which will patch ROMs for you, but it's limited to game gear ROMs (and possibly SMS roms).
In a future version, I may add ROM patching support, as that seems like a natural extention.

## Why is this program better

First, it's written in Java.
It will run on nearly any platform that is of any popularity at all.
Windows, Mac OS X, Linux, Solaris, etc.
The text mode depends upon regular expressions, which were added in Java 1.4, so the text mode does not run on Mac OS classic.
But the applet is written entirely in Java 1.1, so it should.
I don't have a Mac with OS classic, so I cannot test this, but it should theoretically work.

Second, the applet will run in a browser from anywhere in the world on nearly any platform.
You don't even have to download it if you use the GUI version.

Third, there is a text mode AND a GUI mode.
You can script the text mode if you like to do mass-encoding.
Or you can enjoy the convenience of the GUI.

Fourth, it works on every platform the Game Genie was created for.
Nintendo, Super Nintendo, Genesis/Megadrive, Game Boy, and Game Gear.

## What's Missing

Automatic ROM patching would be nice.
But I'll have to learn about SecurityManager's before I could integrate that into the applet.

ggconvc tells you what instruction the hex values correspond to.
I don't know if that's useful to people or not.
It's not useful to me, so I haven't tried to add that.

## How to Use

You will need a Java runtime environment (JRE).
For the text mode, it will need to be Java >= 1.4 because we use regular expressions.
For the GUI mode, any JRE should work, but it's only been tested in Java 1.4.

You can get a Java JRE from Sun. http://java.sun.com

On the command line, you can start the program by typing java -jar ggencoder.jar

This will get you the GUI mode.
To use it in text mode, you'll need to supply some arguments.

`java -jar nintendo decode ATVATGSL` will decode the NES game genie code ATVATGSL into it's raw component parts (60:0466?B5).
This means the value is 60, the address is 0466, and the compare value is B5.
These values are always in hex.

`java -jar nintendo encode 60:0466?B5` will give you the code ATVATGSL.

You can do SNES, Genesis, and Game Boy/Game Gear codes, too.
Just change nintendo to snes, or genesis, or gbgg.

For some more help and examples, type `java -jar ggencoder.jar --help`.

## Credits and Thanks

* Merlyn LeRoy - Process for decypting Genesis/Megadrive codes
* Game Genie Code Creator's Club - Process for decrypting SNES codes
* Benzene - Process for decrypting NES codes
* The Genie's Bottle - Process for decrypting Game Boy / Game Gear Codes
* Nick Joslin? - He may be the author of the document from the Genie's Bottle that I read, but the site is gone, and I no longer have the original document.
* Wyrmcorp - Authors of uggconv. I tested my results against theirs
* Cheatmaster - Author of ggconvc - The whole idea came from her program
* Tony Hedstrom - His Codes gave me the idea to make some of my own, which lead to needing an encoding program, which lead to this
* Johnny Hang Yang - His questions made me take a closer look at the ignored part of NES addresses and lead me to fix a bug in the program.

## How it Works

The source code is here if you want to know how the program works.
If you want to know how the decryptions work, take a look at the docs in the docs directory.

## Problems Questions and Comments

You can contact emuWorks through our [website](http://games.technoplaza.net).
There is an online feedback form if you like, or you can email us at
webmaster ( A T ) technoplaza.net (remove the spaces and change it to @).


ChangeLog
  Version 1.1 - November 16, 2004
	- fixed a bug in NES gg encoder where codes would be invalid on a real gg
  Version 1.0 - July 17, 2004
	- intial release
	- support for NES, SNES, Game Gear, Game Boy, and Genesis Codes


source/ggencoder/GGEncoder.java
/*
 * Copyright (C) 2004 emuWorks
 * http://games.technoplaza.net
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

package ggencoder;

import java.applet.Applet;

import java.awt.Frame;

import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import ggencoder.datastructures.AbstractGameGenieCode;
import ggencoder.datastructures.BasicGBGGRawCode;
import ggencoder.datastructures.BasicGenesisRawCode;
import ggencoder.datastructures.BasicNESRawCode;
import ggencoder.datastructures.BasicSNESRawCode;
import ggencoder.datastructures.GBGGRawCode;
import ggencoder.datastructures.GameGenieCode;
import ggencoder.datastructures.GenesisRawCode;
import ggencoder.datastructures.NESRawCode;
import ggencoder.datastructures.SNESRawCode;

import ggencoder.encoder.AbstractGameGenieDecoder;
import ggencoder.encoder.AbstractGameGenieEncoder;
import ggencoder.encoder.GameGenieDecoder;
import ggencoder.encoder.GameGenieEncoder;

import ggencoder.exceptions.InvalidGameGenieCodeException;

import ggencoder.view.EncoderApplet;

/**
 * The runclass for the game genie encoder program in text or GUI mode.
 *
 * @author John David Ratliff
 * @version 1.4, 11/16/04
 */
public class GGEncoder {
    private static final int SYSTEM_NES = 1;
    private static final int SYSTEM_SNES = 2;
    private static final int SYSTEM_GENESIS = 3;
    private static final int SYSTEM_GBGG = 4;
    private static final int ACTION_DECODE = 1;
    private static final int ACTION_ENCODE = 2;

    /**
     * The application's name.
     */
    public static final String APP_NAME = "Java GGEncoder";

    /**
     * The application's version.
     */
    public static final String APP_VERSION = "1.1";
    
    /**
     * The application's copyright.
     */
    public static final String APP_COPYRIGHT = "Copyright (C) 2004 emuWorks";
    
    /**
     * The application's website.
     */
    public static final String APP_URL = "http://games.technoplaza.net/";

    /**
     * Pattern for recognizing NES Raw Codes.
     */
    public static final Pattern NES_PATTERN = Pattern.compile("^([A-Fa-f0-9]{1,2}):([A-Fa-f0-9]{1,4})(?:\\?([A-Fa-f0-9]{1,2}))?$");

    /**
     * Pattern for recognizing SNES Raw Codes.
     */
    public static final Pattern SNES_PATTERN = Pattern.compile("^([A-Fa-f0-9]{1,2}):([A-Fa-f0-9]{1,6})$");

    /**
     * Pattern for recognizing Genesis Raw Codes.
     */
    public static final Pattern GENESIS_PATTERN = Pattern.compile("^([A-Fa-f0-9]{1,4}):([A-Fa-f0-9]{1,6})$");

    /**
     * Pattern for recognizing GameBoy/GameGear Raw Codes.
     */
    public static final Pattern GBGG_PATTERN = NES_PATTERN;

    /**
     * Creates a new GGEncoder object.
     */
    private GGEncoder() {
        // disallow outside instantiation
        // we should be accessed only from the main method
    }

    /**
     * Displays the copyright notice.
     */
    public static void printCopyright() {
        System.out.println(APP_NAME + " " + APP_VERSION);
        System.out.println("Copyright (C) 2004 emuWorks");
        System.out.println("http://games.technoplaza.net");
    }

    /**
     * Displays syntax and usage information.
     */
    public static void printSyntax() {
        System.out.println("syntax: java -jar ggencoder.jar [ system ] [ action ] [ code [,code2, ... ]]");
        System.out.println("   e.g. java -jar ggencoder.jar or");
        System.out.println("        java -jar ggencoder.jar nintendo decode ATVATGSL");
        System.out.println("        java -jar ggencoder.jar nintendo encode FF:1E51?01 C8:1E52?02");
    }

    /**
     * Displays the help menu.
     */
    public static void printHelp() {
        printCopyright();
        System.out.println();

        printSyntax();
        System.out.println();

        System.out.println("With no arguments, ggencoder starts in GUI mode.");
        System.out.println("system = nintendo | nes | supernintendo | snes | genesis |");
        System.out.println("         megadrive | gameboy | gamegear | gb | gg | gbgg");
        System.out.println("action = decode | encode");
        System.out.println("code   = game genie code, or raw code. Case is ignored.");
        System.out.println("         NES game genie = 6 or 8 letter game genie code");
        System.out.println("         SNES game genie = 9 letter of form XXXX-XXXX");
        System.out.println("         Genesis game genie = 9 letter of form XXXX-XXXX");
        System.out.println("         GameBoy/GameGear game genie = 7 or 11 letter code");
        System.out.println("                                       of form XXX-XXX, or XXX-XXX-XXX");
        System.out.println();
        System.out.println("         NES Raw Code = VV:AAAA (VV = value, AAAA = address)");
        System.out.println("                     or VV:AAAA?CC (CC = compare value)");
        System.out.println("         SNES Raw Code = VV:AAAAAA (VV = value, AAAAAA = address)");
        System.out.println("         Genesis Raw Code = VVVV:AAAAAA (VVVV = value, AAAAAA = address)");
        System.out.println("         GameBoy/GameGear Raw Code = VV:AAAA (VV = value, AAAA = address)");
        System.out.println("                                  or VV:AAAA?CC (CC = compare value)");
        System.out.println("         All values and addresses must be specified in hexadecimal.");
    }

    /**
     * Decodes an NES game genie code.
     *
     * @param code The NES game genie code String.
     */
    public void decodeNES(String code) {
        GameGenieDecoder decoder = AbstractGameGenieDecoder.getInstance();

        try {
            GameGenieCode ggcode = AbstractGameGenieCode.parseNES(code);
            NESRawCode rawcode = decoder.decodeNES(ggcode);

            System.out.println(ggcode + " = " + rawcode);
        } catch (InvalidGameGenieCodeException e) {
            System.err.println(code +
                               " is not a valid NES game genie code.");
        }
    }

    /**
     * Encodes an NES game genie code.
     *
     * @param code The NES raw code String.
     */
    public void encodeNES(String code) {
        GameGenieEncoder encoder = AbstractGameGenieEncoder.getInstance();

        Matcher matcher = NES_PATTERN.matcher(code);

        if (matcher.matches()) {
            int value = Integer.parseInt(matcher.group(1), 16);
            int address = Integer.parseInt(matcher.group(2), 16);
            NESRawCode rawcode;

            if (matcher.group(3) != null) {
                int compare = Integer.parseInt(matcher.group(3), 16);
                rawcode = new BasicNESRawCode(address, value, compare);
            } else {
                rawcode = new BasicNESRawCode(address, value);
            }

            GameGenieCode ggcode = encoder.encode(rawcode);

            System.out.println(rawcode + " = " + ggcode);
        } else {
            System.err.println(code + " is not a valid NES Raw Code.");
            System.err.println("Use the form VV:AAAA?CC, VV=value, AAAA=address, CC=compare (optional)");
            System.err.println("e.g. FF:1E51?01 or 2F:3B19");
        }
    }

    /**
     * Decodes an SNES game genie code.
     *
     * @param code The SNES game genie code String.
     */
    public void decodeSNES(String code) {
        GameGenieDecoder decoder = AbstractGameGenieDecoder.getInstance();

        try {
            GameGenieCode ggcode = AbstractGameGenieCode.parseSNES(code);
            SNESRawCode rawcode = decoder.decodeSNES(ggcode);

            System.out.println(ggcode + " = " + rawcode);
        } catch (InvalidGameGenieCodeException e) {
            System.err.println(code +
                               " is not a valid SNES game genie code.");
        }
    }

    /**
     * Encodes an SNES game genie code.
     *
     * @param code The SNES raw code String.
     */
    public void encodeSNES(String code) {
        GameGenieEncoder encoder = AbstractGameGenieEncoder.getInstance();

        Matcher matcher = SNES_PATTERN.matcher(code);

        if (matcher.matches()) {
            int value = Integer.parseInt(matcher.group(1), 16);
            int address = Integer.parseInt(matcher.group(2), 16);

            SNESRawCode rawcode = new BasicSNESRawCode(address, value);
            GameGenieCode ggcode = encoder.encode(rawcode);

            System.out.println(rawcode + " = " + ggcode);
        } else {
            System.err.println(code + " is not a valid SNES Raw Code.");
            System.err.println("Use the form VV:AAAAAA, VV=value, AAAAAA=address");
            System.err.println("e.g. 44:01AF62");
        }
    }

    /**
     * Decodes a Genesis game genie code.
     *
     * @param code The Genesis game genie code String.
     */
    public void decodeGenesis(String code) {
        GameGenieDecoder decoder = AbstractGameGenieDecoder.getInstance();

        try {
            GameGenieCode ggcode = AbstractGameGenieCode.parseGenesis(code);
            GenesisRawCode rawcode = decoder.decodeGenesis(ggcode);

            System.out.println(ggcode + " = " + rawcode);
        } catch (InvalidGameGenieCodeException e) {
            System.err.println(code +
                               " is not a valid Genesis game genie code.");
        }
    }

    /**
     * Encodes a Genesis game genie code.
     *
     * @param code The Genesis raw code String.
     */
    public void encodeGenesis(String code) {
        GameGenieEncoder encoder = AbstractGameGenieEncoder.getInstance();

        Matcher matcher = GENESIS_PATTERN.matcher(code);

        if (matcher.matches()) {
            int value = Integer.parseInt(matcher.group(1), 16);
            int address = Integer.parseInt(matcher.group(2), 16);

            GenesisRawCode rawcode = new BasicGenesisRawCode(address, value);
            GameGenieCode ggcode = encoder.encode(rawcode);

            System.out.println(rawcode + " = " + ggcode);
        } else {
            System.err.println(code + " is not a valid Genesis Raw Code.");
            System.err.println("Use the form VVVV:AAAAAA, VVVV=value, AAAAAA=address");
            System.err.println("e.g. 44:01AF62");
        }
    }

    /**
     * Decodes a GameBoy/GameGear game genie code.
     *
     * @param code The GameBoy/GameGear game genie code String.
     */
    public void decodeGBGG(String code) {
        GameGenieDecoder decoder = AbstractGameGenieDecoder.getInstance();

        try {
            GameGenieCode ggcode = AbstractGameGenieCode.parseGBGG(code);
            GBGGRawCode rawcode = decoder.decodeGBGG(ggcode);

            System.out.println(ggcode + " = " + rawcode);
        } catch (InvalidGameGenieCodeException e) {
            System.err.println(code +
                               " is not a valid GameBoy/GameGear game genie code.");
        }
    }

    /**
     * Encodes a GameBoy/GameGear game genie code.
     *
     * @param code The GameBoy/GameGear raw code String.
     */
    public void encodeGBGG(String code) {
        GameGenieEncoder encoder = AbstractGameGenieEncoder.getInstance();

        Matcher matcher = GBGG_PATTERN.matcher(code);

        if (matcher.matches()) {
            int value = Integer.parseInt(matcher.group(1), 16);
            int address = Integer.parseInt(matcher.group(2), 16);
            GBGGRawCode rawcode;

            if (matcher.group(3) != null) {
                int compare = Integer.parseInt(matcher.group(3), 16);
                rawcode = new BasicGBGGRawCode(address, value, compare);
            } else {
                rawcode = new BasicGBGGRawCode(address, value);
            }

            GameGenieCode ggcode = encoder.encode(rawcode);

            System.out.println(rawcode + " = " + ggcode);
        } else {
            System.err.println(code + " is not a valid GBGG Raw Code.");
            System.err.println("Use the form VV:AAAA?CC, VV=value, AAAA=address, CC=compare (optional)");
            System.err.println("e.g. 00:4A17?C8 or 2A:2E36");
        }
    }

    /**
     * Runs the program in TEXT mode.
     *
     * @param args The program arguments.
     */
    public void startTextMode(String[] args) {
        int system = 0;
        int action = 0;

        String arg = args[0];

        if ((arg.compareToIgnoreCase("nes") == 0) ||
            (arg.compareToIgnoreCase("nintendo") == 0)) {
            system = SYSTEM_NES;
        } else if ((arg.compareToIgnoreCase("snes") == 0) ||
                   (arg.compareToIgnoreCase("supernintendo") == 0)) {
            system = SYSTEM_SNES;
        } else if ((arg.compareToIgnoreCase("genesis") == 0) ||
                   (arg.compareToIgnoreCase("megadrive") == 0)) {
            system = SYSTEM_GENESIS;
        } else if ((arg.compareToIgnoreCase("gameboy") == 0) ||
                   (arg.compareToIgnoreCase("gamegear") == 0) ||
                   (arg.compareToIgnoreCase("gb") == 0) ||
                   (arg.compareToIgnoreCase("gg") == 0) ||
                   (arg.compareToIgnoreCase("gbgg") == 0)) {
            system = SYSTEM_GBGG;
        } else {
            printSyntax();
            System.exit(-1);
        }

        arg = args[1];

        if (arg.compareToIgnoreCase("decode") == 0) {
            action = ACTION_DECODE;
        } else if (arg.compareToIgnoreCase("encode") == 0) {
            action = ACTION_ENCODE;
        } else {
            printSyntax();
            System.exit(-1);
        }

        Method method = null;
        Class[] argTypes = new Class[1];
        argTypes[0] = String.class;

        try {
            switch (system) {
                case SYSTEM_NES:

                    if (action == ACTION_DECODE) {
                        method = this.getClass().getMethod("decodeNES",
                                                           argTypes);
                    } else {
                        method = this.getClass().getMethod("encodeNES",
                                                           argTypes);
                    }

                    break;
                case SYSTEM_SNES:

                    if (action == ACTION_DECODE) {
                        method = this.getClass().getMethod("decodeSNES",
                                                           argTypes);
                    } else {
                        method = this.getClass().getMethod("encodeSNES",
                                                           argTypes);
                    }

                    break;
                case SYSTEM_GENESIS:

                    if (action == ACTION_DECODE) {
                        method = this.getClass().getMethod("decodeGenesis",
                                                           argTypes);
                    } else {
                        method = this.getClass().getMethod("encodeGenesis",
                                                           argTypes);
                    }

                    break;
                case SYSTEM_GBGG:

                    if (action == ACTION_DECODE) {
                        method = this.getClass().getMethod("decodeGBGG",
                                                           argTypes);
                    } else {
                        method = this.getClass().getMethod("encodeGBGG",
                                                           argTypes);
                    }

                    break;
            }
        } catch (NoSuchMethodException e) {
            // this should be impossible
            e.printStackTrace();
            System.exit(-1);
        }

        for (int i = 2; i < args.length; i++) {
            Object[] passedArgs = new Object[1];
            passedArgs[0] = args[i];

            try {
                method.invoke(this, passedArgs);
            } catch (IllegalAccessException e) {
                // this should be impossible
                e.printStackTrace();
            } catch (InvocationTargetException e) {
                e.printStackTrace();
            }
        }
    }

    /**
     * Called to start the program.
     *
     * @param args The program arguments. Use --help for information.
     */
    public static void main(String[] args) {
        if (args.length == 0) {
            Frame frame = new Frame(APP_NAME + " " + APP_VERSION);

            Applet applet = new EncoderApplet();
            applet.init();
            applet.start();

            frame.add(applet);
            frame.setSize(400, 350);
            frame.addWindowListener(new WindowAdapter() {
                    public void windowClosing(WindowEvent e) {
                        System.exit(0);
                    }
                });
            frame.setVisible(true);
        } else if (args.length == 1) {
            if (args[0].compareToIgnoreCase("--help") == 0) {
                printHelp();
            } else {
                printSyntax();
            }
        } else if (args.length >= 3) {
            new GGEncoder().startTextMode(args);
        } else {
            printSyntax();
        }
    }
}
